In [1]:
# Krina Bharatbhai Patel (NetID : vj3004)
# Vatsal Pathak (NetID : pv2556)
# Group ID : 02

import numpy as np
from pandas import DataFrame, read_csv, to_numeric
from sklearn import cluster
from sklearn.model_selection import train_test_split
from sklearn.naive_bayes import GaussianNB
from sklearn import metrics
from sklearn.metrics import accuracy_score
from sklearn.metrics import mean_squared_error
from sklearn.linear_model import LogisticRegression
from sklearn.neighbors import KNeighborsClassifier

In [2]:
data = read_csv('diabetes.csv', sep=",")
data.columns

Index(['Pregnancies', 'Glucose', 'BloodPressure', 'SkinThickness', 'Insulin',
       'BMI', 'DiabetesPedigreeFunction', 'Age', 'Outcome'],
      dtype='object')

In [3]:
x = data.drop(columns=['Outcome'])
y = data['Outcome']

In [4]:
XT = ['XT01', 'XT02', 'XT03', 'XT04', 'XT05', 'XT06', 'XT07', 'XT08', 'XT09', 'XT10' ]
XV = ['XV01', 'XV02', 'XV03', 'XV04', 'XV05', 'XV06', 'XV07', 'XV08', 'XV09', 'XV10']

In [5]:
for i in range(1,11):
    if i < 10:
        temp_t = "XT"+str(0)+str(i)
        temp_v = "XV"+str(0)+str(i)
    else:
        temp_t = "XT"+str(i)
        temp_v = "XV"+str(i)
        
    data_train = read_csv('diabetes_10fold_train_val/'+temp_t+'.csv')
    data_train_x = data_train.drop(columns=['Outcome'])
    data_train_y = data_train['Outcome']
    data_val = read_csv('diabetes_10fold_train_val/'+temp_v+'.csv')
    data_val_x = data_val.drop(columns=['Outcome'])
    data_val_y = data_val['Outcome']
    
    nbmodel = GaussianNB()
    nbmodel.fit(data_train_x,data_train_y)
    nbpredictions_train = nbmodel.predict(data_train_x)
    nbMSE_train = mean_squared_error(data_train_y,nbpredictions_train)
    print('Naive Bayes Classifier: Fold - ' + str(i) + ' Training Error : ' '%.4f' % (nbMSE_train) )
    
    nbpredictions_val = nbmodel.predict(data_val_x)
    nbMSE_val = mean_squared_error(data_val_y,nbpredictions_val)
    print('Naive Bayes Classifier: Fold - ' + str(i) + ' Validation Error : ' '%.4f' % (nbMSE_val) )
    
    lrmodel = LogisticRegression(solver='lbfgs', max_iter=1000)
    lrmodel.fit(data_train_x,data_train_y)
    lrpredictions_train = lrmodel.predict(data_train_x)
    lrMSE_train = mean_squared_error(data_train_y,lrpredictions_train)
    print('Logistic Regression Classifier: Fold - ' + str(i) + ' Training Error : ' '%.4f' % (lrMSE_train) )
    
    lrpredictions_val = lrmodel.predict(data_val_x)
    lrMSE_val = mean_squared_error(data_val_y,lrpredictions_val)
    print('Logistic Regression Classifier: Fold - ' + str(i) + ' Validation Error : ' '%.4f' % (lrMSE_val) )
  

    knnmodel = KNeighborsClassifier(n_neighbors=3)
    knnmodel.fit(data_train_x,data_train_y)
    knnpredictions_train = knnmodel.predict(data_train_x)
    knnMSE_train = mean_squared_error(data_train_y,knnpredictions_train)
    print('K-Nearest Neighbors Classifier: Fold - ' + str(i) + ' Training Error : ' '%.4f' % (knnMSE_train) )
    
    knnpredictions_val = knnmodel.predict(data_val_x)
    knnMSE_val = mean_squared_error(data_val_y,knnpredictions_val)
    print('K-Nearest Neighbors Classifier: Fold - ' + str(i) + ' Validation Error : ' '%.4f' % (knnMSE_val) )
    
    
    
    

Naive Bayes Classifier: Fold - 1 Training Error : 0.2370
Naive Bayes Classifier: Fold - 1 Validation Error : 0.2105
Logistic Regression Classifier: Fold - 1 Training Error : 0.2182
Logistic Regression Classifier: Fold - 1 Validation Error : 0.2105
K-Nearest Neighbors Classifier: Fold - 1 Training Error : 0.1488
K-Nearest Neighbors Classifier: Fold - 1 Validation Error : 0.2895
Naive Bayes Classifier: Fold - 2 Training Error : 0.2344
Naive Bayes Classifier: Fold - 2 Validation Error : 0.2727
Logistic Regression Classifier: Fold - 2 Training Error : 0.2214
Logistic Regression Classifier: Fold - 2 Validation Error : 0.2338
K-Nearest Neighbors Classifier: Fold - 2 Training Error : 0.1462
K-Nearest Neighbors Classifier: Fold - 2 Validation Error : 0.4286
Naive Bayes Classifier: Fold - 3 Training Error : 0.2330
Naive Bayes Classifier: Fold - 3 Validation Error : 0.2727
Logistic Regression Classifier: Fold - 3 Training Error : 0.2127
Logistic Regression Classifier: Fold - 3 Validation Error :

# Training Error
| Fold | Naïve Bayes classifier | Logistic Regression classifier | 3-NN classifier |
|------|------|------|------|
|1|0.2370|0.2182|0.1488|
|2|0.2344|0.2214 |0.1462 |
|3|0.2330|0.2127 |0.1418 |
|4|0.2475|0.2272 |0.1577 |
|5|0.2446|0.2200 |0.1360 |
|6|0.2359|0.2171 |0.1520 |
|7|0.2402|0.2200 |0.1621 |
|8|0.2301|0.2171 |0.1389 |
|9|0.2489|0.2214 |0.1433 |
|10|0.2254|0.2124 |0.1402 |


# Validation Error
| Fold | Naïve Bayes classifier | Logistic Regression classifier | 3-NN classifier |
|------|------|------|------|
|1|0.2105|0.2105|0.2895|
|2|0.2727|0.2338 |0.4286 |
|3|0.2727|0.2987 |0.2468 |
|4|0.1818|0.1299 |0.2078 |
|5|0.2338|0.2078 |0.2987 |
|6|0.2727|0.2208 |0.3247 |
|7|0.2468|0.2208 |0.2597 |
|8|0.2338|0.2597 |0.3247 |
|9|0.1688|0.1688 |0.3117 |
|10|0.3158|0.2895 |0.4079 |